password_enhancements.module

```diff
@@ -0,0 +1,484 @@
+  // Non-unique constrains from the same policy won't override each other,
+  // although if a higher priority policy's constraint defines that specific
+  // type then it will override each non-unique constraint from the lower
+  // priority from the same type.
+  $checked_constraints = [];
+  $constraint_list = [];
+  foreach ($policies as $policy) {
+    $usable_constraints_by_policy = [];
+
+    foreach ($constraints as $constraint_id => $constraint) {
+      if ($constraint->getPolicy() == $policy->id()) {
+        $type = $constraint->getType();
+
+        if (!array_key_exists($type, $checked_constraints)) {
+          $usable_constraints_by_policy[$type] = TRUE;
+          $constraint_list[$constraint_id] = $constraint;
+        }
+
+        // Remove already checked constraint.
+        unset($constraints[$constraint_id]);
+      }
+    }
+
+    $checked_constraints += $usable_constraints_by_policy;
+  }
+
+  return $constraint_list;
+}
+
+/**
+ * Ajax callback for updating the constraints on role change.
+ *
+ * @param array $form
+ *   The form.
+ * @param \Drupal\Core\Form\FormStateInterface $form_state
+ *   The current form state.
+ *
+ * @return array
+ *   Renderable form array for password policy constraints.
+ */
+function password_enhancements_update_constraints_by_roles(array $form, FormStateInterface $form_state): array {
+  return $form['account']['password_policy_constraints'];
+}
+
+/**
+ * Implements hook_entity_base_field_info().
+ */
+function password_enhancements_entity_base_field_info(EntityTypeInterface $entity_type): array {
+  $fields = [];
+
+  if ($entity_type->id() == 'user') {
+    $fields['password_enhancements_password_changed_date'] = BaseFieldDefinition::create('datetime')
+      ->setLabel(t('Last password change'))
+      ->setDescription(t('The last time when the user has changed the password.'))
+      ->setDefaultValue(0);
+
+    $fields['password_enhancements_password_change_required'] = BaseFieldDefinition::create('boolean')
+      ->setLabel(t('Password change required'))
+      ->setDescription(t('Forces the user to change the password.<br><strong>Warning:</strong> This also causes to force logout the user.'))
+      ->setDefaultValue(0)
+      ->setDisplayOptions('form', [
+        'type' => 'boolean_checkbox',
+        'settings' => [
+          'display_label' => TRUE,
+        ],
+        'weight' => 99,
+      ]);
+  }
+
+  return $fields;
+}
+
+/**
+ * Implements hook_ENTITY_TYPE_presave().
+ */
+function password_enhancements_user_presave(EntityInterface $entity) {
+  // Update the last password changed date if the password was changed.
+  if ($entity->bundle() == 'user' && (empty($entity->original) || $entity->original->pass->value != $entity->pass->value)) {
+    $entity->password_enhancements_password_changed_date->setValue(Drupal::time()->getRequestTime());
+  }
+}
+
+/**
+ * Implements hook_user_login().
+ */
+function password_enhancements_user_login(UserInterface $user): void {
```

Comment:
Don't you want to auto-logout a user when it is logged in but its password expires? Like the automated logout module does.

My gut says what you do here should be handled in an event subscriber. For example in a `\Symfony\Component\HttpKernel\KernelEvents::REQUEST` subscriber.

Reply:
Let's mark this as resolved for now.